RecoTracker/TkSeedGenerator/plugins/SeedingOTEDProducer.cc

```diff
@@ -0,0 +1,380 @@
+              buildingPropagator->setPropagationDirection(oppositeToMomentum);
+            } else if (searchingPropagator->propagationDirection() == oppositeToMomentum) {
+              buildingPropagator->setPropagationDirection(alongMomentum);
+            }
+
+            updatedTSOSL1.second.rescaleError(100);
+
+            TrajectoryStateOnSurface updatedTSOSL1_final = updator_->update(updatedTSOSL1.second, *hitL1);
+            if UNLIKELY (!updatedTSOSL1_final.isValid())
+              continue;
+            std::pair<bool, TrajectoryStateOnSurface> updatedTSOSL2_final =
+                propagateAndUpdate(updatedTSOSL1_final, *buildingPropagator, *hitL2);
+            if (!updatedTSOSL2_final.first)
+              continue;
+            std::pair<bool, TrajectoryStateOnSurface> updatedTSOSL3_final =
+                propagateAndUpdate(updatedTSOSL2_final.second, *buildingPropagator, hitL3);
+            if (!updatedTSOSL3_final.first)
+              continue;
+            TrajectorySeed ts =
+                createSeed(updatedTSOSL3_final.second, container, hitL3.geographicalId(), *buildingPropagator);
+            result.push_back(ts);
+          }
+        }
+      }
+    }
+  }
+  result.shrink_to_fit();
+  return result;
+}
+
+unsigned int SeedingOTEDProducer::checkLayer(unsigned int iidd) {
+  StripSubdetector strip = StripSubdetector(iidd);
+  unsigned int subid = strip.subdetId();
+  if (subid == StripSubdetector::TIB || subid == StripSubdetector::TOB) {
+    return tkTopo_->layer(iidd);
+  }
+  return 0;
+}
+
+std::vector<VectorHit> SeedingOTEDProducer::collectVHsOnLayer(const edmNew::DetSetVector<VectorHit>& input,
+                                                              unsigned int layerNumber) {
+  std::vector<VectorHit> VHsOnLayer;
+  if (!input.empty()) {
+    for (const auto& DSViter : input) {
+      if (checkLayer(DSViter.id()) == layerNumber) {
+        for (const auto& vh : DSViter) {
+          VHsOnLayer.push_back(vh);
+        }
+      }
+    }
+  }
+
+  return VHsOnLayer;
+}
+
+void SeedingOTEDProducer::printVHsOnLayer(const edmNew::DetSetVector<VectorHit>& input, unsigned int layerNumber) {
+  if (!input.empty()) {
+    for (const auto& DSViter : input) {
+      for (const auto& vh : DSViter) {
+        if (checkLayer(DSViter.id()) == layerNumber)
+          LogTrace("SeedingOTEDProducer") << " VH in layer " << layerNumber << " >> " << vh;
+      }
+    }
+  } else {
+    LogTrace("SeedingOTEDProducer") << " No VHs in layer " << layerNumber << ".";
+  }
+}
+
+const TrajectoryStateOnSurface SeedingOTEDProducer::buildInitialTSOS(const VectorHit& vHit) const {
+  // having fun with theta
+  Global3DVector gv(vHit.globalPosition().x(), vHit.globalPosition().y(), vHit.globalPosition().z());
+  float theta = gv.theta();
+  // gv transform to local (lv)
+  const Local3DVector lv(vHit.det()->surface().toLocal(gv));
+
+  //FIXME::charge is fine 1 every two times!!
```

Comment:
I must admit that this isn't clear to me either. Maybe @ebrondol can shed some light on it, otherwise I will implement the negative charge being tried as well.

Reply:
@JanFSchulte yes, please go ahead and change it. It doesnt make much sense that try only one charge. Thanks!